In [18]:
import warnings

warnings.filterwarnings("ignore")

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel\ipkernel.py:287: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [2]:
import pandas as pd
import numpy as np

In [3]:
#!pip install --user mlxtend

In [4]:
from mlxtend.frequent_patterns import apriori,association_rules
from mlxtend.preprocessing import TransactionEncoder

In [19]:
data = pd.read_csv("D:\YUKTA\Assignment\DS\Association Rules\Online retail.csv")

In [20]:
data

Transactions
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7498                                            chicken
7499                                 escalope,green tea
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7501 rows x 1 columns]

In [21]:
data.shape

(7501, 1)

In [22]:
df = list(data["Transactions"].apply(lambda x:x.split(",") ))

Transforming the list with one-hot encoding:

In [33]:
a = TransactionEncoder()

In [34]:
a_data = a.fit(df).transform(df)

In [35]:
df = pd.DataFrame(a_data,columns=a.columns_)
df = df.replace(False,0)
df = df.replace(True,1)

In [36]:
df

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0              0        1                  1          0        1            0   
1              0        0                  0          0        0            0   
2              0        0                  0          0        0            0   
3              0        0                  0          0        1            0   
4              0        0                  0          0        0            0   
...          ...      ...                ...        ...      ...          ...   
7496           0        0                  0          0        0            0   
7497           0        0                  0          0        0            0   
7498           0        0                  0          0        0            0   
7499           0        0                  0          0        0            0   
7500           0        0                  0          0        0            0   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0         0               0          0            0  ...       0   
1         0               0          0            0  ...       0   
2         0               0          0            0  ...       0   
3         0               0          0            0  ...       1   
4         0               0          0            0  ...       0   
...     ...             ...        ...          ...  ...     ...   
7496      0               0          0            0  ...       0   
7497      0               0          0            0  ...       0   
7498      0               0          0            0  ...       0   
7499      0               0          0            0  ...       0   
7500      0               0          0            0  ...       0   

      vegetables mix  water spray  white wine  whole weat flour  \
0                  1            0           0                 1   
1                  0            0           0                 0   
2                  0            0           0                 0   
3                  0            0           0                 0   
4                  0            0           0                 0   
...              ...          ...         ...               ...   
7496               0            0           0                 0   
7497               0            0           0                 0   
7498               0            0           0                 0   
7499               0            0           0                 0   
7500               0            0           0                 0   

      whole wheat pasta  whole wheat rice  yams  yogurt cake  zucchini  
0                     0                 0     1            0         0  
1                     0                 0     0            0         0  
2                     0                 0     0            0         0  
3                     0                 0     0            0         0  
4                     0                 1     0            0         0  
...                 ...               ...   ...          ...       ...  
7496                  0                 0     0            0         0  
7497                  0                 0     0            0         0  
7498                  0                 0     0            0         0  
7499                  0                 0     0            0         0  
7500                  0                 0     0            1         0  

[7501 rows x 120 columns]

Products that customers bought or did not buy during shopping will are represented by values 1 and 0 respectively.

# Apriori Algorithm

In [37]:
df_apriori = apriori(df, min_support = 0.05, use_colnames = True)
df_apriori

support                    itemsets
0   0.087188                   (burgers)
1   0.081056                      (cake)
2   0.059992                   (chicken)
3   0.163845                 (chocolate)
4   0.080389                   (cookies)
5   0.051060               (cooking oil)
6   0.179709                      (eggs)
7   0.079323                  (escalope)
8   0.170911              (french fries)
9   0.063325           (frozen smoothie)
10  0.095321         (frozen vegetables)
11  0.052393             (grated cheese)
12  0.132116                 (green tea)
13  0.098254               (ground beef)
14  0.076523            (low fat yogurt)
15  0.129583                      (milk)
16  0.238368             (mineral water)
17  0.065858                 (olive oil)
18  0.095054                  (pancakes)
19  0.071457                    (shrimp)
20  0.050527                      (soup)
21  0.174110                 (spaghetti)
22  0.068391                  (tomatoes)
23  0.062525                    (turkey)
24  0.058526          (whole wheat rice)
25  0.052660  (chocolate, mineral water)
26  0.050927       (mineral water, eggs)
27  0.059725  (spaghetti, mineral water)

# Conclusion:

From the above table we can se that **Mineral Water** is the most frequently purchased item with a support of 0.2383.

It indicates that approximately 23.84% of transactions include mineral water.

**Eggs** and **French Fries** are also highly popular, with supports of 0.179709 and 0.170911, respectively.

**Spaghetti** and **Chocolate** follow closely, with supports of 0.174110 and 0.163845, respectively.

Items like **Soup (0.050527)**, **Cooking Oil (0.051060)**, and **Grated Cheese (0.052393)** are among the less frequently purchased items, each appearing in just over 5% of transactions.

# Association Rules

In [38]:
rules = association_rules(df_apriori, metric="lift", min_threshold=0.5)

In [39]:
rules

antecedents      consequents  antecedent support  consequent support  \
0      (chocolate)  (mineral water)            0.163845            0.238368   
1  (mineral water)      (chocolate)            0.238368            0.163845   
2  (mineral water)           (eggs)            0.238368            0.179709   
3           (eggs)  (mineral water)            0.179709            0.238368   
4      (spaghetti)  (mineral water)            0.174110            0.238368   
5  (mineral water)      (spaghetti)            0.238368            0.174110   

    support  confidence      lift  leverage  conviction  zhangs_metric  
0  0.052660    0.321400  1.348332  0.013604    1.122357       0.308965  
1  0.052660    0.220917  1.348332  0.013604    1.073256       0.339197  
2  0.050927    0.213647  1.188845  0.008090    1.043158       0.208562  
3  0.050927    0.283383  1.188845  0.008090    1.062815       0.193648  
4  0.059725    0.343032  1.439085  0.018223    1.159314       0.369437  
5  0.059725    0.250559  1.439085  0.018223    1.102008       0.400606

In [40]:
rules.sort_values('lift',ascending = False).head(10)

antecedents      consequents  antecedent support  consequent support  \
4      (spaghetti)  (mineral water)            0.174110            0.238368   
5  (mineral water)      (spaghetti)            0.238368            0.174110   
0      (chocolate)  (mineral water)            0.163845            0.238368   
1  (mineral water)      (chocolate)            0.238368            0.163845   
2  (mineral water)           (eggs)            0.238368            0.179709   
3           (eggs)  (mineral water)            0.179709            0.238368   

    support  confidence      lift  leverage  conviction  zhangs_metric  
4  0.059725    0.343032  1.439085  0.018223    1.159314       0.369437  
5  0.059725    0.250559  1.439085  0.018223    1.102008       0.400606  
0  0.052660    0.321400  1.348332  0.013604    1.122357       0.308965  
1  0.052660    0.220917  1.348332  0.013604    1.073256       0.339197  
2  0.050927    0.213647  1.188845  0.008090    1.043158       0.208562  
3  0.050927    0.283383  1.188845  0.008090    1.062815       0.193648

# Conclusion:

There is a strong association between Chocolate and Mineral Water.

The rule **(chocolate -> mineral water)** has a confidence of 32.14% and a lift of 1.35. 

This indicates that customers who buy chocolate are 32.14% likely to also buy mineral water.

The rule **(mineral water -> eggs)** has a confidence of 21.36% and a lift of 1.19.

While the association is not as strong as the chocolate and mineral water combination, it still indicates that customers who buy mineral water are 1.19 times more likely to also buy eggs than expected by chance.

There is also a strong association between Mineral Water and Spaghetti.

The rule **(mineral water -> spaghetti)** has a confidence of 25.06% and a lift of 1.44, indicating a significant association.

Thus showing that customers buying mineral water are 1.44 times more likely to buy spaghetti compared to random chance.